[Network, Security] Bindings for Network, Security and some classes needed to port samples

migueldeicaza opened the pull request:
Test build

spouliot commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+			}
+		}
+
+		[Mac(10,14)][iOS(12,0)][Watch(5,0)][TV(12,0)]
```
Sadly the framework is not available on watchOS so all `[Watch(5,0)]` should be removed.

migueldeicaza replied:
It is indeed, the header files show that everywhere :-(

spouliot commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+		static extern string nw_endpoint_get_bonjour_service_type (OS_nw_endpoint endpoint);
+
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		public string BonjourSericeType => nw_endpoint_get_bonjour_service_type (handle);
```
typo `Serice`

monojenkins commented:
Build failure
Build comment file: 
--------------
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7368//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7368//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7368//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7368//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7368//Generator_Diff)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7368//Test_Report/) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

--------------

rolfbjarne commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+		// extern const struct sockaddr * _Nonnull nw_endpoint_get_address (nw_endpoint_t _Nonnull endpoint) __attribute__((availability(tvos, introduced=12.0))) __attribute__((availability(watchos, introduced=5.0))) __attribute__((availability(ios, introduced=12.0))) __attribute__((availability(macos, introduced=10.14)));
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern IntPtr /* struct sockaddr* */ nw_endpoint_get_address (OS_nw_endpoint endpoint);
```
This isn't used/exposed, in which case it should probably be commented out.

migueldeicaza replied:
That is a work in progress

rolfbjarne commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+			if (serviceType == null)
+				throw new ArgumentNullException (nameof (serviceType));
+			var x = nw_endpoint_create_bonjour_service (name, serviceType, domain);
+			return new NWEndpoint (x, owns: true);
```
`nw_endpoint_create_bonjour_service` can return NULL.

rolfbjarne commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+	
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern unsafe string nw_endpoint_get_bonjour_service_name (OS_nw_endpoint endpoint);
```
This P/Invoke will free the string returned from `nw_endpoint_get_bonjour_service_name`, but it's defined as `const char *` in the header, so it shouldn't be freed.

migueldeicaza replied:
good catch!

rolfbjarne commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+		
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern string nw_endpoint_get_bonjour_service_type (OS_nw_endpoint endpoint);
```
Same: this P/Invoke will free the returned string.

rolfbjarne commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,163 @@
+		
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern string nw_endpoint_get_bonjour_service_domain (OS_nw_endpoint endpoint);
```
Same: this P/Invoke will free the returned string.

monojenkins commented:
Build failure
Build comment file: 
--------------
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7370//console)
🔥 [Build failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7370//console) 🔥

--------------

rolfbjarne commented on src/Network/NWParameters.cs:
```diff
@@ -0,0 +1,148 @@
+	
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern void nw_parameters_set_prefer_no_proxy (nw_parameters_t parameters, bool prefer_no_proxy);
```
I think all bools in P/Invokes need a `MarshalAs(UnmanagedType.I1)` attribute, since the native `bool` type is 1 byte, and by default they're 4 bytes in P/Invokes.

monojenkins commented:
Build failure
Build comment file: 
--------------
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7371//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7371//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7371//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7371//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7371//Generator_Diff)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7371//Test_Report/) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

--------------

migueldeicaza commented:
Status: https://gist.github.com/migueldeicaza/08491ba75fe3037f6a54c01ae6788c12

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7516//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7516//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7516//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7516//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7516//Generator_Diff)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7516//Test_Report/) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7520//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7520//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7520//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7520//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7520//Generator_Diff)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7520//Test_Report/) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7521//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7521//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7521//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7521//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7521//Generator_Diff)

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7522//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7522//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7522//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7522//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7522//Generator_Diff)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7522//Test_Report/) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7542//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7542//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7542//API_diff_(from_stable))
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7542//API_diff_(PR_only))
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7542//Generator_Diff)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7542//Test_Report/) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
🔥 [Build failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7564//console) 🔥

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7566//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7566//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7566//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7566//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7566//Test_20Report) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7571//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7571//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7571//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7571//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7571//Test_20Report) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7576//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7576//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7576//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7576//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7576//Test_20Report) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7598//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7598//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7598//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7598//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7598//Test_20Report) 🔥

# Test results

9 tests failed, 55 tests passed.

## Failed tests

 * apitest/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * apitest/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * dont link/Mac Unified XM45: Failed (Test run crashed (exit code: 134).)
 * introspection/Mac Unified: Failed (Test run crashed (exit code: 134).)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed

monojenkins commented:
Build failure

!!! Couldn't read commit file !!!

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7631//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7631//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7631//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7631//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7631//Test_20Report) 🔥

# Test results

6 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 4)

migueldeicaza commented:
Work-in-progress netcat port:

https://gist.github.com/migueldeicaza/100e944240e5264d90c6a5452de97b74

monojenkins commented:
Build success
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7632//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7632//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7632//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7632//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7632//Test_20Report) 🔥

# Test results

6 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 4)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7642//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7642//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7642//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7642//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7642//Test_20Report) 🔥

# Test results

6 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 4)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7654//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7654//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7654//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7654//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7654//Test_20Report) 🔥

# Test results

6 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 4)

monojenkins commented:
Build success
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7655//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7655//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7655//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7655//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7655//Test_20Report) 🔥

# Test results

6 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 4)

monojenkins commented:
Build failure
❌ [Build was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7662/)

monojenkins commented:
Build failure
❌ [Build was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7663/)

🔥 [Build failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7663//console) 🔥

monojenkins commented:
Build failure
❌ [Build was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7666/)

🔥 [Build failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7666//console) 🔥

monojenkins commented:
Build failure
❌ [Build was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7682/)

🔥 [Build failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7682//console) 🔥

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7689//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7689//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7689//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7689//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7689//Test_20Report) 🔥

# Test results

7 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * introspection/watchOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 5)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7696//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7696//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7696//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7696//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7696//Test_20Report) 🔥

# Test results

7 tests failed, 60 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * introspection/watchOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 5)

migueldeicaza commented:
NetCat sample is now here, what i use to try this out and explore the API:

https://github.com/migueldeicaza/NetCatNetwork

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7721//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7721//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7721//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7721//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7721//Test_20Report) 🔥

# Test results

6 tests failed, 61 tests passed.

## Failed tests

 * introspection/Mac Unified: Failed (Test run failed.)
 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 13)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7747//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7747//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7747//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7747//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7747//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 5)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7759//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7759//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7759//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7759//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7759//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 2)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7772//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7772//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7772//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7772//Generator_20Diff) (please review changes)

rolfbjarne commented:
build

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7783//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7783//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7783//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7783//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7783//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: BuildFailure
 * introspection/iOS Unified 64-bits - simulator/Debug: BuildFailure
 * introspection/tvOS - simulator/Debug: BuildFailure
 * MTouch tests/NUnit: Failed (Execution failed with exit code 2)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7789//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7789//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7789//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7789//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7789//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 2)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7801//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7801//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7801//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7801//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7801//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: BuildFailure
 * introspection/iOS Unified 64-bits - simulator/Debug: BuildFailure
 * introspection/tvOS - simulator/Debug: BuildFailure
 * MTouch tests/NUnit: Failed (Execution failed with exit code 2)

monojenkins commented:
Build failure
❌ [Build was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7813/)

✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7813//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7813//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7813//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7813//Generator_20Diff) (please review changes)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7817//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7817//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7817//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7817//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7817//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 3)

rolfbjarne commented on src/Network/NWContentContext.cs:
```diff
@@ -0,0 +1,232 @@
+		public static NWContentContext DefaultStream {
+			get {
+				if (defaultStream == null)
+					defaultStream = MakeGlobal (Marshal.ReadIntPtr (Dlfcn.dlsym (Libraries.Network.Handle, "_nw_content_context_default_stream")));
```
Why do you need to make these objects global when they're stored in static variables anyway?

migueldeicaza replied:
To prevent someone calling `Dispose()` and ruining the global instance for everyone else.

spouliot commented on src/Security/SecCertificate2.cs:
```diff
@@ -0,0 +1,46 @@
+	public class SecCertificate2 : NativeObject {
+		public SecCertificate2 (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
```
not needed in members (when identical to the type-level attributes)

spouliot commented on src/Security/SecIdentity2.cs:
```diff
@@ -0,0 +1,84 @@
+			if (identity == null)
+				throw new ArgumentNullException (nameof (identity));
+			
+			handle = sec_identity_create (identity == null ? IntPtr.Zero : identity.Handle);
```
That API can return `null` so `InitializeHandle` (from base `NativeObject`) should be called to avoid creating an invalid instance (that will crash/throw later).

migueldeicaza replied:
Got it - will use the property version.

spouliot commented on src/Security/SecIdentity2.cs:
```diff
@@ -0,0 +1,84 @@
+			if (certificates == null)
+				throw new ArgumentNullException (nameof (certificates));
+			using (var nsarray = NSArray.FromObjects (certificates)){
+				handle = sec_identity_create_with_certificates (identity.Handle, nsarray.Handle);
```
same -> `InitializeHandle`

migueldeicaza replied:
Got it - will use the property version.

spouliot commented on src/Security/SecIdentity2.cs:
```diff
@@ -0,0 +1,84 @@
+				for (int i = 0; i < n; i++){
+					ret [i] = new SecCertificate (NSArray.GetAtIndex (certArray, (nuint) i), owns: false);
+				}
+				return ret;
```
`certArray` should be disposed before returning

migueldeicaza replied:
got it

spouliot commented on src/Security/SecProtocolMetadata.cs:
```diff
@@ -0,0 +1,69 @@
+		// if this ever changes, make this public
+		internal SecProtocolMetadata (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
```
same, move to type-level

migueldeicaza replied:
got it

spouliot commented on src/Security/SecProtocolOptions.cs:
```diff
@@ -0,0 +1,203 @@
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.SecurityLibrary)]
+		static extern void sec_protocol_options_set_challenge_block(sec_protocol_options_t options, IntPtr challenge_block, dispatch_queue_t challenge_queue);
```
better comment this (and the next one), since they are not publicly exposed. That way xtro can report them as missing/ignored

migueldeicaza replied:
got it

spouliot commented on src/Security/SecTrust2.cs:
```diff
@@ -0,0 +1,48 @@
+			if (trust == null)
+				throw new ArgumentNullException (nameof (trust));
+			
+			handle = sec_trust_create (trust == null ? IntPtr.Zero : trust.Handle);
```
can return `null`, should use `InitializeHandle`

spouliot commented on src/generator-attributes.cs:
```diff
@@ -856,3 +856,14 @@ public DefaultEnumValueAttribute ()
+// the side effects of the binding of the method, rather than the
+// method itself -- for security for example, we generate the
+// proxy code for invoking blocks and callbacks to blocks
+// see 
```
incomplete comment ?

spouliot commented on tests/introspection/ApiCMAttachmentTest.cs:
```diff
@@ -416,9 +425,45 @@ protected INativeObject GetINativeInstance (Type t)
 					new CVPixelBufferAttributes (CVPixelFormatType.CV24RGB, 100, 50)
 				);
+			case "NWAdvertiseDescriptor":
+				return NWAdvertiseDescriptor.CreateBonjourService ("sampleName" + DateTime.Now);
```
The NRE in the tests is because `GetINativeInstance` returns `null` which is likely because of the extra entries being added returns `null`. It could be this one

spouliot commented on tests/introspection/ApiCMAttachmentTest.cs:
```diff
@@ -416,9 +425,45 @@ protected INativeObject GetINativeInstance (Type t)
+			case "NWContentContext":
+				return new NWContentContext ("contentContext" + DateTime.Now);
+			case "NWEndpoint":
+				return NWEndpoint.Create ("www.microsoft.com", "https");
```
this one

spouliot commented on tests/introspection/ApiCMAttachmentTest.cs:
```diff
@@ -416,9 +425,45 @@ protected INativeObject GetINativeInstance (Type t)
+				// Makes a new instance every time
+				return NWProtocolDefinition.TcpDefinition;
+			case "NWProtocolOptions":
+				return NWProtocolOptions.CreateTcp ();
```
or one of the above...

migueldeicaza commented on tools/common/Frameworks.cs:
```diff
@@ -410,6 +412,7 @@ public static Frameworks GetwatchOSFrameworks (Application app)
 					{ "Vision", "Vision", 11 },
 
 					{ "NaturalLanguage", "NaturalLanguage", 12,0 },
+					{ "Network", "Network", 14, 0 } ,
```
Got it

migueldeicaza commented on src/CoreFoundation/DispatchData.cs:
```diff
@@ -0,0 +1,134 @@
+				throw new ArgumentException (nameof (start));
+			if (length < 0)
+				throw new ArgumentException (nameof (length));
+			if (start+length >= buffer.Length)
```
Replaced with:

```
			if (start > buffer.Length + length)
```

migueldeicaza commented on src/CoreFoundation/DispatchIO.cs:
```diff
@@ -0,0 +1,117 @@
+// Copyright 2010 Novell, Inc.
+// Copyright 2011-2014 Xamarin Inc
+//
+// this contains DispathcObject, Group and Queue
```
got it

migueldeicaza commented on src/Network/NWAdvertiseDescriptor.cs:
```diff
@@ -0,0 +1,66 @@
+		public NWAdvertiseDescriptor (IntPtr handle, bool owns) : base (handle, owns)
+		{ }
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
```
got it

migueldeicaza commented on src/Network/NWAdvertiseDescriptor.cs:
```diff
@@ -0,0 +1,66 @@
+		{
+			if (txt == null)
+				throw new ArgumentNullException (nameof (txt));
+			var n = System.Text.Encoding.UTF8.GetByteCount (txt);
```
Because Mono marshals `System.String` using UTF8 encoding, and this API needs the number of bytes being passed.

migueldeicaza commented on src/Network/NWConnection.cs:
```diff
@@ -0,0 +1,574 @@
+	public class NWConnection : NativeObject {
+		public NWConnection (IntPtr handle, bool owns) : base (handle, owns) {} 
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
```
got it

migueldeicaza commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,133 @@
+	public class NWEndpoint : NativeObject {
+		public NWEndpoint (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[Mac(10,14)][iOS(12,0)][TV(12,0)]
```
got it

migueldeicaza commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,133 @@
+		static extern IntPtr nw_endpoint_get_hostname (OS_nw_endpoint endpoint);
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		public string Hostname => Marshal.PtrToStringAnsi (nw_endpoint_get_hostname (GetHandle()));
```
This one returns a pointer to an internal string, the other one returns a copy.

PInvoke with a string signature releases return values.

migueldeicaza commented on src/Network/NWEndpoint.cs:
```diff
@@ -0,0 +1,133 @@
+		[DllImport (Constants.NetworkLibrary)]
+		static extern IntPtr /* struct sockaddr* */ nw_endpoint_get_address (OS_nw_endpoint endpoint);
+
+		// TODO: same
```
We want to, but we need to have a discussion about what to do and how to surface it.   This is going to be needed, but we can not use the .NET data type, we are going to need a new data type.

migueldeicaza commented on src/Network/NWInterface.cs:
```diff
@@ -0,0 +1,49 @@
+	public class NWInterface : NativeObject {
+		public NWInterface (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
```
got it

migueldeicaza commented on src/Network/NWListener.cs:
```diff
@@ -0,0 +1,253 @@
+			if (parameters == null)
+				throw new ArgumentNullException (nameof (parameters));
+			if (port == null)
+				throw new ArgumentNullException (nameof (parameters));
```
got it

migueldeicaza commented on src/Network/NWListener.cs:
```diff
@@ -0,0 +1,253 @@
+			if (del != null){
+				var nwendpoint = new NWEndpoint (endpoint, owns: false);
+			        del (nwendpoint, added != 0 ? true : false);
+				nwendpoint.Dispose ();
```
Oh thank god.   This bit me hard, and ended up adding fences to track down this bug.

migueldeicaza replied:
Updating all the sites that had this problem.

migueldeicaza replied:
Thinking about this a little more - I think that the send/receive stacks probably should document instead that the instance is not usable after the callback, as the GetINativeObject path is slower than wrapping the instance.

migueldeicaza commented on src/Network/NWParameters.cs:
```diff
@@ -0,0 +1,516 @@
+		static public class Network {
+                        static public readonly IntPtr Handle = Dlfcn.dlopen (Constants.NetworkLibrary, 0);
+                }
+	}
```
It is not being generated, I do not know the reason for this.

migueldeicaza commented on src/Network/NWParameters.cs:
```diff
@@ -0,0 +1,516 @@
+                        if (del != null){
+				var x = new NWProtocolOptions (iface, owns: false);
+				del (x);
+				x.Dispose ();
```
handled

migueldeicaza commented on src/Network/NWParameters.cs:
```diff
@@ -0,0 +1,516 @@
+                        if (del != null){
+				var x = new NWInterface (iface, owns: false);
+                                var ret = del (x);
+				x.Dispose ();
```
handled

migueldeicaza commented on src/Network/NWPath.cs:
```diff
@@ -0,0 +1,153 @@
+	public class NWPath : NativeObject {
+		public NWPath (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
```
got it

migueldeicaza commented on src/Network/NWPath.cs:
```diff
@@ -0,0 +1,153 @@
+		extern static bool nw_path_has_ipv4 (IntPtr handle);
+		
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		public bool HasIpV4 => nw_path_has_ipv4 (GetHandle());
```
got i

migueldeicaza commented on src/Network/NWPath.cs:
```diff
@@ -0,0 +1,153 @@
+		extern static bool nw_path_has_ipv6 (IntPtr handle);
+		
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		public bool HasIpV6 => nw_path_has_ipv6 (GetHandle());
```
got it

migueldeicaza commented on src/Network/NWPath.cs:
```diff
@@ -0,0 +1,153 @@
+			var descriptor = (BlockLiteral *) block;
+			var del = (Action<NWInterface>) (descriptor->Target);
+			if (del != null){
+				del (new NWInterface (iface, owns: false));
```
got it

migueldeicaza commented:
Addressed most of the issues here, but two caveats:

* I do not follow your guidance on the null issue being raised, need some help.

* Did not remove the CoreGraphics code, will merge that here.

* I suspect that we do not want GetINativeObject in the send/receive main methods, and rather document that you are not supposed to keep the object, as it will be released (and the check is already in place).      So that would be the NWError for Send, and the NWContentContext/NWError in Receive.

We should revisit the latter.

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7847//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7847//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7847//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7847//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7847//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 1)

mandel-macaque commented:
The failing tests is due to the NWAdvertiseDescriptor.CreateBonjourService returning a null pointer which means that later accessing the handler returns all the errors seen in the report. The fix is quite simple, the method must take the type. A number of example types that Mac OS X uses can be found [here](https://developer.apple.com/library/archive/qa/qa1312/_index.html).

For example, changing the test to use the following fixes it:
```csharp
return NWAdvertiseDescriptor.CreateBonjourService ("sampleName" + DateTime.Now, "_nfs._tcp");
```
But this makes me wonder if we should not make type to be optional, the documentation on the headers only states that domain is optional, but not type (will add a review in the PR). But as you can see in the header docs:

```
*!
+ * @function nw_advertise_descriptor_create_bonjour_service
+ *
+ * @abstract
+ *		Creates a new advertise descriptor object based on a Bonjour service type
+ *		and optional domain. This object can be used with listener objects to
+ *      specify the service the listener should advertise.
+ *
+ *      If the name is unspecified, the default name for the device will be used.
+ *      If the domain is unspecified, the default domains for registration will be
+ *      used. If the listener is local only, the domain 'local.' will be used
+ *      regardless of the parameter passed to domain.
+ *
+ * @param name
+ *		An optional Bonjour service name.
+ *
+ * @param type
+ *		A Bonjour service type.
+ *
+ * @param domain
+ *		An optional Bonjour service domain.
+ *
+ * @result
+ *		An instantiated browse descriptor object.
+ */
```

mandel-macaque commented on src/Network/NWAdvertiseDescriptor.cs:
```diff
@@ -0,0 +1,62 @@
+		[DllImport (Constants.NetworkLibrary)]
+		static extern IntPtr nw_advertise_descriptor_create_bonjour_service (string name, string type, string domain);
+
+		public static NWAdvertiseDescriptor CreateBonjourService (string name, string type = null, string domain = null)
```
@migueldeicaza I would not add type to be optional, since that will most (if not always) result in the nw_advertise_descriptor_create_bonjour_service returning a null pointer. I added a comment on how to fix the issue in the tests.

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7857//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7857//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7857//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7857//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7857//Test_20Report) 🔥

# Test results

5 tests failed, 62 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure
 * introspection/iOS Unified 32-bits - simulator/Debug: Failed
 * introspection/iOS Unified 64-bits - simulator/Debug: Failed
 * introspection/tvOS - simulator/Debug: Failed
 * MTouch tests/NUnit: Failed (Execution failed with exit code 1)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7867//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7867//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7867//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7867//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7867//Test_20Report) 🔥

# Test results

2 tests failed, 65 tests passed.

## Failed tests

 * Xtro/Mac: Failed (Test run crashed (exit code: 32).)
 * MTouch tests/NUnit: Failed (Execution failed with exit code 1)

spouliot commented on src/Security/SecProtocolOptions.cs:
```diff
@@ -0,0 +1,201 @@
+		static extern void sec_protocol_options_set_tls_min_version (sec_protocol_options_t handle, SslProtocol protocol);
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		void SetTlsMinVersion (SslProtocol protocol) => sec_protocol_options_set_tls_min_version (GetCheckedHandle (), protocol);
```
`public` ?

migueldeicaza replied:
agreed.

spouliot commented on src/Security/SecProtocolOptions.cs:
```diff
@@ -0,0 +1,201 @@
+		static extern void sec_protocol_options_set_tls_max_version (sec_protocol_options_t handle, SslProtocol protocol);
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		void SetTlsMaxVersion (SslProtocol protocol) => sec_protocol_options_set_tls_max_version (GetCheckedHandle (), protocol);
```
`public` ?

migueldeicaza replied:
agreed.

spouliot commented on src/Security/SecProtocolOptions.cs:
```diff
@@ -0,0 +1,201 @@
+		static extern void sec_protocol_options_add_tls_application_protocol (sec_protocol_options_t handle, string applicationProtocol);
+
+		[TV (12,0), Mac (10,14), iOS (12,0)]
+		void AddTlsApplicationProtocol (string applicationProtocol)
```
`public` ?

migueldeicaza replied:
agreed.

monojenkins commented:
Build failure
❌ [Build was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7882/)

monojenkins commented:
Build failure
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7885//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7885//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7885//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7885//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7885//Test_20Report) 🔥

# Test results

1 tests failed, 66 tests passed.

## Failed tests

 * Xtro/Mac: BuildFailure

monojenkins commented:
Build success
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7887//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7887//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7887//API_20diff_20_28PR_20only_29) (please review changes)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7887//Generator_20Diff) (please review changes)
✅ [Test run succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/7887//Test_20Report)

migueldeicaza commented:
The source code for CFRetain shows that it handles NSObject retain as well:

https://opensource.apple.com/source/CF/CF-550/CFRuntime.c

Mike Ash also says that this is allowed:

https://www.mikeash.com/pyblog/friday-qa-2010-01-22-toll-free-bridging-internals.html

This shows that when it is dealing with an NSObject, this invokes the ObjC retain/release.

That said, I do not mind changing that to call NSObject.retain/NSObject release

migueldeicaza commented:
Hello - and I just validated that calling CFRelease/CFRetain on an NSObject does indeed call the overwritten method.
